File: 백준/Silver/2606. 바이러스/바이러스.py
# 1번 컴터랑 연결된 컴터들이 바이러스에 걸려서 총4대 도루묵
# 컴퓨터 갯수에 따른 위치 배치 --> 우린 보통 인덱스 1~7을 사용하지 인덱스 0은 버리걸랑
# 연결된 링크는 6개  (출발 s -> 끝 e      끝 e -> 출발 s)
# 연결된 링크 --> 마치 인접행렬?

# 첫 위치를 방문표시하고 바로 시작한다

# 2. dfs
def dfs(c):
    global ans
    ans += 1

    v[c] = 1

    for n in adj[c]:
        if not v[n]:
            dfs(n)

# 1.입출력 셋팅
import sys
input = sys.stdin.readline
N= int(input())
M = int(input())

# 컴퓨터 세팅 --> 인덱스 0~7, 근데 실제로는 1~7만 필요
# 링크 방문 노드 세팅
adj = [ [] for _ in range(N+1)]
v = [0] * (N+1)

for _ in range(M):
    s, e = list(map(int, input().split()))
    adj[s].append(e)
    adj[e].append(s)


ans = 0
dfs(1)
print(ans-1)

File: 백준/Silver/2606. 바이러스/바이러스.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;

// graph
import java.util.ArrayList;

// DFS
import java.util.Stack;
import java.util.List;

// sys
import java.util.Arrays;
import java.util.Collections;


public class Main {
	static ArrayList<Integer>[] graph;
	static boolean[] visited;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		int M = Integer.parseInt(br.readLine());
		
		// allocating the size of graph
		graph = new ArrayList[N+1];
		for (int i = 1; i < N+1; i++) {
			graph[i] = new ArrayList<>();
		}
		
		for (int i = 0; i < M; i++) {
			String[] ab = br.readLine().split(" ");
			int a = Integer.parseInt(ab[0]);
			int b = Integer.parseInt(ab[1]);
			graph[a].add(b);
			graph[b].add(a);
		}
		
		// asc
		for (int i = 1; i < N+1; i++) {
			Collections.sort(graph[i]);
		}
		
		
		// res
		int V = 1;
		visited = new boolean[N+1];
		DFS(V);
	}
	
	
	
	
	public static void DFS(int start) {
		Stack<Integer> stack = new Stack<>();
		stack.push(start);
		
		int cnt = 0;
		while (!stack.isEmpty()) {
			int node = stack.pop();
			
			
			if (!visited[node]) visited[node] = true;
			else if (visited[node]) continue;
//			System.out.print(node + " ");
			cnt++;
			
			// neighbors
			List<Integer> neighbor = graph[node];
			for (int i = neighbor.size()-1; i >= 0; i--) {
				int next = neighbor.get(i);
				if (!visited[next]) {
					stack.push(next);
				}
			}
		}
		System.out.println(--cnt);
		
	}
	
	
	
}